core/platform/email/gae_email_services_test.py

```diff
@@ -23,23 +23,310 @@
+        self.assertEqual(len(message_b), 0)
+        with self.swap(feconf, 'CAN_SEND_EMAILS', True):
+            gae_email_services.send_bulk_mail(
+                self.SENDER_EMAIL, self.RECIPIENT_EMAILS,
+                'subject', 'body', 'html')
+        message_a = self.mail_stub.get_sent_messages(
+            to=self.RECIPIENT_EMAILS[0])
+        self.assertEqual(len(message_a), 1)
+        message_b = self.mail_stub.get_sent_messages(
+            to=self.RECIPIENT_EMAILS[1])
+        self.assertEqual(len(message_b), 1)
+
+    def test_bulk_emails_not_sent_if_can_send_emails_is_false(self):
+        # Emails are not sent if the CAN_SEND_EMAILS setting is not turned on.
+        message_a = self.mail_stub.get_sent_messages(
+            to=self.RECIPIENT_EMAILS[0])
+        self.assertEqual(len(message_a), 0)
+        message_b = self.mail_stub.get_sent_messages(
+            to=self.RECIPIENT_EMAILS[1])
+        self.assertEqual(len(message_b), 0)
+        email_exception = self.assertRaisesRegexp(
+            Exception, 'This app cannot send emails.')
+        with email_exception:
+            gae_email_services.send_bulk_mail(
+                self.SENDER_EMAIL, self.RECIPIENT_EMAILS,
+                'subject', 'body', 'html')
+        message_a = self.mail_stub.get_sent_messages(
+            to=self.RECIPIENT_EMAILS[0])
+        self.assertEqual(len(message_a), 0)
+        message_b = self.mail_stub.get_sent_messages(
+            to=self.RECIPIENT_EMAILS[1])
+        self.assertEqual(len(message_b), 0)
+
+    def test_bulk_mails_not_sent_if_sender_email_is_malformed(self):
+        message_a = self.mail_stub.get_sent_messages(
+            to=self.RECIPIENT_EMAILS[0])
+        self.assertEqual(0, len(message_a))
+        message_b = self.mail_stub.get_sent_messages(
+            to=self.RECIPIENT_EMAILS[1])
+        self.assertEqual(0, len(message_b))
+        malformed_sender_email = ''
+        email_exception = self.assertRaisesRegexp(
+            ValueError, 'Malformed sender email address: %s'
+            % malformed_sender_email)
+        with self.swap(feconf, 'CAN_SEND_EMAILS', True), email_exception:
+            gae_email_services.send_bulk_mail(
+                malformed_sender_email, self.RECIPIENT_EMAILS,
+                'subject', 'body', 'html')
+        message_a = self.mail_stub.get_sent_messages(
+            to=self.RECIPIENT_EMAILS[0])
+        self.assertEqual(0, len(message_a))
+        message_b = self.mail_stub.get_sent_messages(
+            to=self.RECIPIENT_EMAILS[1])
+        self.assertEqual(0, len(message_b))
+
+        message_a = self.mail_stub.get_sent_messages(
+            to=self.RECIPIENT_EMAILS[0])
+        self.assertEqual(0, len(message_a))
+        message_b = self.mail_stub.get_sent_messages(
+            to=self.RECIPIENT_EMAILS[1])
+        self.assertEqual(0, len(message_b))
+        malformed_sender_email = 'malformed_sender_email@example.com'
```

Comment:
Ditto (re providing context for malformed-ness).

Reply:
Done